src/mscorlib/shared/System/Buffers/MemoryManager.cs

```diff
@@ -0,0 +1,66 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Runtime;
+using System.Runtime.CompilerServices;
+
+namespace System.Buffers
+{
+    /// <summary>
+    /// Manager of Memory<typeparamref name="T"/> that provides the implementation.
+    /// </summary>
+    public abstract class MemoryManager<T> : IMemoryOwner<T>, IPinnable
+    {
+        /// <summary>
+        /// The number of items in the Memory<typeparamref name="T"/>.
+        /// </summary>
+        public virtual int Length => GetSpan().Length;
```

Comment:
With this change; any reason for this to be virtual?

Reply:
Probably to avoid the `GetSpan()` call?

Reply:
Could have done
```csharp
public int Length { get; }

protected MemoryManager(int length)
{
    Length = length;
}

public Memory<T> Memory => new Memory<T>(this, 0, Length);
```
Unless its valid that derived classes of `MemoryManager` can change in length on the go?

Reply:
@benaadams There are scenarios where this can change; e.g., if the factory pools and reuses these instances.

Reply:
Why would it change in case of a pool? We usually don't return exact size requested.

Reply:
For fixed sizes (general pooling and all the corefx, aspnet uses) then they become inlinable properties so you get nice perf optimization automatically.

To take advantage of changing sizes the pool needs to be doing defragmentation and some pretty advanced memory management; while not being able to move anything currently allocated unlike the GC can. Is that a likely scenario? 

Is easier to add more chunks and have multiple block sized pools?

Reply:
Also you could achieve it with?
```csharp
public int Length { get; protected set; }
```

Reply:
I have no idea why Length was ever on OwnedMemory...

Reply:
Is two things to get out of sync. 

Either `Length` and `Memory` should be non virtual with a protected .ctor (best pref, hardest to get wrong)

Or `Length` shouldn't exist and `Memory` should be abstract (non-inlinable, but again less to get wrong)

Reply:
> Either `Length` and `Memory` should be non virtual with a protected .ctor
```csharp
public class MyMemory : MemoryManager<byte>
{
    public MyMemory(...) : base(length)
    {
        // ...
    }
} 
```
Are these implementations ever going to be something different? Why make them virtual?
```csharp
public int Length { get; /* protected set; */ }
public Memory<T> Memory => new Memory<T>(this, 0, Length);
```

Reply:
```csharp
public class MyResizableMemory : MemoryManager<byte>
{
    public MyMemory(...) : base(length)
    {
        // ...
    }

    public int Size { get => Length; set => Length = value; }
} 
```

Reply:
Perhaps need an Offset also
```csharp
public int Length { get; }
protected int Offset { get; set; }

protected MemoryManager(int length)
{
    Length = length;
}

public Memory<T> Memory => new Memory<T>(this, Offset, Length);
```

Atm the virtual calls cannot be inlined; and the default implementation for getting a `Memory` from a `MemoryManager` is terrible:

`non-inline Memory`
 -> `non-inline Length`
 -> `non-inline GetSpan`
 ->` validation`
 -> `Length`
 -> `Memory .ctor`

 so you are forcing people to implement a method for no reason; when they could be POCO items rather than virtual and all inline just fine:

inlinable `Memory .ctor`

Reply:
@benaadams @GrabYourPitchforks just made me aware of the fact that this type will rarely be used directly because the MemoryPool returns `IMemoryOwner<T>`, so really none of this matters. Though as an implementer of `MemoryManager<T>`, I think it's confusing to expose length. So lets try to come up with something concrete.

Reply:
K so `Memory` won't inline due to the interface. 

However as the implementer you will have to get a `Length` somewhere and possibly an `Offset`, so can stick it on the type and have everything else inline and have a good performance implementation with no additional virtuals or the need to implement the two methods?

```csharp
public abstract class MemoryManager<T> : IMemoryOwner<T>, IPinnable
{
    protected int Offset { get; set; }
    protected int Length { get; set; }

    protected MemoryManager()
    {
    }
    
    protected MemoryManager(int length)
    {
        Length = length;
    }

    protected MemoryManager(int offset, int length)
    {
        Offset = offset;
        Length = length;
    }

    public Memory<T> Memory => new Memory<T>(this, Offset, Length);
    
    // Rest as now
```

Reply:
Or kill `Length` and force the implementation
```csharp
public abstract class MemoryManager<T> : IMemoryOwner<T>, IPinnable
{
    public abstract Memory<T> Memory { get; }
    
    // Rest as now
```

Reply:
@GrabYourPitchforks making it abstract?

Reply:
Thanks for the feedback. In the interest of time, I am going to merge the change as is and we can consider changing this separately.